Migrate to jgrapht

paagamelo opened the pull request:
My own graph implementation was dropped in favour of `jgrapht` library, interfaces were removed as well a part from `NavigationGraph`.
Additionally, generic `Obstacle`s and `EnvironmentWithObstacle`s were modelled.

paagamelo commented on alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/AbstractOrientingBehavior.kt:
```diff
@@ -183,7 +180,7 @@ abstract class AbstractOrientingBehavior<T, A, N, E, M, F>(
             nextRoom
         } else {
-            environment.graph().nodes().first { it.contains(currentPosition) }
+            environment.graph().vertexSet().first { it.contains(currentPosition) }
```
It's explicitly stated in the `NavigationGraph` interface that each implementation must guarantee predictable ordering for the collections it maintains. It's part of the contract defined by that interface

paagamelo commented on alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/geometry/euclidean/twod/navigator/NaviGator.kt:
```diff
@@ -83,22 +83,22 @@ fun generateNavigationGraph(
-                val crossings =
-                    seed.findCrossings(index, seeds, origin, width, height, obstacles, unity)
+                val passages =
+                    seed.findPassages(index, seeds, origin, width, height, obstacles, unity)
```
Probably I thought it was, but I changed my mind now, fixed.

paagamelo commented on alchemist/alchemist-implementationbase/src/main/java/it/unibo/alchemist/model/implementations/utils/RectObstacle2D.java:
```diff
@@ -24,11 +25,11 @@
  */
 @SuppressFBWarnings("EQ_DOESNT_OVERRIDE_EQUALS")
-public final class RectObstacle2D extends Rectangle2D.Double implements Obstacle2D {
+public final class RectObstacle2D extends Rectangle2D.Double implements Obstacle2D<Euclidean2DPosition> {
```
The bound of that type parameter is actually a `Position2D`, I have thought quite a lot about whether to parameterize `Obstacle` on a `Vector` or a `Position`. The latter is best when using obstacles in environments (which are parameterised on positions as well) and allows not to break older code.  Additionally, I believe that obstacles may be meaningful even not in vector spaces (think e.g. of prohibited airspace).
Given that, if you agree to maintain the `Position` bound, `RectObstacle2D` is clearly meant to be placed in a bidimensional vector space, thus should only work with `Euclidean2DPosition` (at least to the best of my knowledge, which is quite limited when it comes to geometry)

DanySK replied:
@paagamelo the thing is conceptual: what do you need out of an obstacle? If you do vectorial operations (sum, subtraction, multiplication) then you necessarily need vectors, not positions.

DanySK commented on alchemist/alchemist-swingui/src/main/java/it/unibo/alchemist/boundary/gui/effects/DrawNavigationGraph.java:
```diff
@@ -106,7 +106,7 @@
             });
         }
-        if (env instanceof ImageEnvironmentWithGraph) {
+        if (graph == null && env instanceof ImageEnvironmentWithGraph) {
```
Why is this on the bottom of the method and not on the beginning? Doesn't it skip one update?

paagamelo replied:
It did, fixed.

DanySK commented on alchemist/src/orchid/resources/wiki/usage/navigation-graphs.md:
```diff
@@ -22,7 +23,7 @@ The algorithm implemented in Alchemist is called NaviGator (Navigation Graphs ge
 If your environment is codified as an image (such as the planimetry showed above), generating a navigation graph is straight-forward. All you have to do is mark the areas of the environment where to plant initial seeds in blue (RGB #0000FF). In the image below you can see the generation of the navigation graph showed on the top of the page. The blue regions in the original image indicate where to ...
 
-![alt text](../../assets/media/usage/navigation-graph-generation.jpeg)
+![navigaiton graph generation]({{ 'assets/media/usage/navigation-graph-generation.jpeg'|asset }})
```
navigation

DanySK commented on alchemist/src/orchid/resources/wiki/usage/navigation-graphs.md:
```diff
@@ -5,7 +5,8 @@ title: Generating navigation graphs
 A navigation graph of an environment is a graph whose nodes are convex shapes representing portions of the environment which are traversable by agents (namely, walkable areas), and edges represent connections between them. The image below shows a bidimensional environment with obstacles on the left and the associated navigation graph on the right (nodes are painted blue whereas edges are represen ...
 
-![alt text](../../assets/media/usage/navigation-graph.jpeg)
+![navigaiton graph]({{ 'assets/media/usage/navigation-graph.jpeg'|asset }})
```
navigation

DanySK commented on alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/graph/GraphExtension.kt:
```diff
@@ -0,0 +1,26 @@
+ */
+fun <V> Graph<V, *>.pathExists(source: V, sink: V): Boolean =
+    /*
+     * [DijkstraShortestPath] is used instead of [org.jgrapht.alg.connectivity.ConnectivityInspector.pathExists],
```
hey move this to the heading Kdoc

paagamelo commented on alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/cognitiveagents/characteristics/cognitive/OrientingAgent.kt:
```diff
@@ -18,7 +17,7 @@ interface OrientingAgent<
     A : GeometricTransformation<V>,
     N : ConvexGeometricShape<V, A>,
-    E : GraphEdge<N>
+    E
```
Fixed. However, I thought you preferred to use `where` only when class declaration is "heavy" (e.g. in `OrientingBehavior2D`, which is the example you sent me on Teams). But if you prefer to use the where clause everywhere, I'll change the declaration of some other classes you missed, so as to have general coherence.

DanySK replied:
I am still trying to understand myself which style looks better. My current rule of thumbs is that if it crosses the line length, then probably `where` is better. But sometimes is better also if there are non trivial bounds like `A : B<out C, D<in P>>`, or multiple bounds on different types.
On-site declaration looks better for trivial bounds, such as nullability prevention with `A : Any`

paagamelo replied:
Ok, I'll follow this practice

paagamelo commented on alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/graph/NavigationGraphImpl.kt:
```diff
@@ -1,23 +1,95 @@
+ * A directed [Euclidean2DNavigationGraph].
+ */
+typealias DirectedEuclidean2DNavigationGraph =
+    DirectedNavigationGraph<Euclidean2DPosition, Euclidean2DTransformation, ConvexPolygon, Euclidean2DPassage>
```
In this case the `typealias` is syntactic sugar, there are no extension functions.

paagamelo commented:
@DanySK I changed the parameterization of `Obstacle` from `Position` to `Vector`. The interface `Obstacle` is actually pretty general and could describe obstacles not in vector spaces. However, the interface `Obstacle2D` needs vectors as you said, so I could either make `Obstacle` work with vectors (which is what I did, and this means obstacles are now bound to vector spaces, and environments with obstacles as well), or add another interface defining an obstacle in a vector space (which seemed overkill to me at least for the moment).

I modified `RectObstacle2D` so as to work with `Vector2D` instead of `Euclidean2DPosition` only. In doing this, I had to suppress an unchecked cast warning in `Continuous2DObstacles.readObject`. I think what I did is correct, but I couldn't understand why that warning was generated only after my changes, and not before. Maybe check it just to make sure everything's ok.

DanySK commented:
@paagamelo point me to the file/line

paagamelo commented on alchemist/alchemist-implementationbase/src/main/java/it/unibo/alchemist/model/implementations/environments/Continuous2DObstacles.java:
```diff
@@ -150,12 +162,13 @@ private void writeObject(final ObjectOutputStream o) throws IOException {
         o.writeObject(getObstacles());
     }
 
+    @SuppressWarnings("unchecked")
```
@DanySK here

DanySK replied:
What if you cast to RectObstacle2D<?>?

paagamelo replied:
It complains because `rtree` should contain `RectObstacle2D<Euclidean2DPosition>`

DanySK replied:
Meh, leave the suppression there then

paagamelo commented on alchemist/alchemist-interfaces/src/main/kotlin/it/unibo/alchemist/model/interfaces/Obstacle.kt:
```diff
@@ -0,0 +1,26 @@
+ *
+ * @param V the vector type for the space in which this obstacle is placed.
+ */
+interface Obstacle<V : Vector<V>> : Serializable {
```
In the first place, I moved all the methods of `Obstacle2D` to `Obstacle` (and of course I used the type parameter). However, I later changed my mind and moved all those methods back to `Obstacle2D`, but kept the type parameter on `Obstacle`. At present, there is no particular reason to maintain it, even though removing it means `EnvironmentWithObstacle` will become an environment with generic obstacles (instead of a vector space with coherent obstacles).

DanySK replied:
Let's reason for a moment on what an `Obstacle` is then. As it stands, an obstacle is "something with an identifier". What do we need out of an Obstacle?

paagamelo replied:
An obstacle is an n-dimensional shape providing some specific methods related to the intersection with vectors, which are those used by environments with obstacles. These method are now defined in `Obstacle2D`, I'll move them back to `Obstacle`.

DanySK replied:
Makes sense
